src/solvers/refinement/string_refinement_util.h

```diff
@@ -0,0 +1,315 @@
+  /// Record expressions that are contained in the equation with the given index
+  std::unordered_map<std::size_t, std::vector<exprt>> strings_in_equation;
+};
+
+/// Base class for string primitives
+class string_primitivet
+{
+public:
+  string_primitivet() = default;
+  string_primitivet(const string_primitivet &) = delete;
+
+  virtual optionalt<array_string_exprt> string_result() const
+  {
+    return {};
+  }
+
+  virtual std::vector<array_string_exprt> string_arguments() const
+  {
+    return {};
+  }
+};
+
+/// String primitive transforming one string into another
+class string_transformation_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  array_string_exprt input;
+  std::vector<exprt> args;
+  exprt return_code;
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return {input};
+  }
+};
+
+/// String inserting a string into another one
+class string_insertion_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  array_string_exprt input1;
+  array_string_exprt input2;
+  std::vector<exprt> args;
+  exprt return_code;
+
+  /// Constructor from arguments of a function application
+  string_insertion_primitivet(
+    const std::vector<exprt> &fun_args,
+    array_poolt &array_pool);
+
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return {input1, input2};
+  }
+};
+
+/// String creation from other types
+class string_creation_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  std::vector<exprt> args;
+  exprt return_code;
+
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+};
+
+/// String test
+class string_test_primitivet : public string_primitivet
+{
+public:
+  exprt result;
+  std::vector<array_string_exprt> under_test;
+  std::vector<exprt> args;
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return under_test;
+  }
+};
+
+/// Keep track of dependencies between strings.
+/// Each string points to primitive calls on which it depends,
+/// each primitive points to the strings on which the result depend.
+class dependency_grapht
+{
+public:
+  /// A primitive node is just an index in the `primitive_nodes` vector.
+  class primitive_nodet
+  {
+  public:
+    std::size_t index;
+    explicit primitive_nodet(std::size_t i) : index(i)
+    {
+    }
+  };
+
+  /// A string node points to primitive on which it depends
+  class string_nodet
+  {
+  public:
+    std::vector<primitive_nodet> dependencies;
+
+    // In case it depends on a primitive we don't support yet
+    bool depends_on_unknown_primitive = false;
+  };
+
+  string_nodet &get_node(const array_string_exprt &e);
```

Comment:
`get_node` and `make_node` seem like bad names to distinguish the two node types

Reply:
They are different in behaviour in that `make_node` always create a new node, whereas `get_node` first looks at whether it exists or not, otherwise I would have given both function the same name.